### ISSUE TITLE
scArches：原理、reference model构建与迁移学习

### ISSUE THREAD

**Issue by ixxmu** [ixxmu/mp_duty#7338]
https://mp.weixin.qq.com/s/wdR5AxPa30LS_HctobO5CQ

**Comment by ixxmu** [comment]
scArches：原理、reference model构建与迁移学习 by 生信菜鸟团
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/Go1sLc9TF8UQMjMUHzSxWDXlyI5KSFZs5icNBCoWslBAEK1icybHz2HbRcjHDfTvBTNhNia4gySnebAGqaCbpfQduEIoiaKmBIH7Lr0zxVBIpek/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.2722222222222222" data-type="png" data-w="1080" data-imgfileid="100056497" data-aistatus="1" src="https://mmbiz.qpic.cn/mmbiz_png/Go1sLc9TF8UQMjMUHzSxWDXlyI5KSFZs5icNBCoWslBAEK1icybHz2HbRcjHDfTvBTNhNia4gySnebAGqaCbpfQduEIoiaKmBIH7Lr0zxVBIpek/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">scArches 是一种面向 reference mapping 的迁移学习策略，核心目标是在已有 reference atlas 的基础上，将新的 query 数据映射到同一个 latent space 中，从而支持后续的比较分析、标签迁移与自动注释。在 </span><code><span leaf="">scvi-tools</span></code><span leaf=""> 中，这套流程主要围绕 </span><code><span leaf="">SCVI</span></code><span leaf=""> 和 </span><code><span leaf="">SCANVI</span></code><span leaf=""> 展开。</span></p><h2 data-tool="mdnice编辑器"><span><span leaf="">scArches原理</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">首先是 reference 阶段：使用多个公开的 reference datasets，这些数据集可能来自不同 study、不同 batch、不同技术平台，甚至也可以对应不同器官、物种或疾病状态。用条件变分自编码器来训练这些 reference 数据，并给每个数据集一个条件标签（Reference labels）。训练完成之后，模型会得到一个共享的 latent space，在这个空间里，batch、technology 这类条件效应被尽量回归掉，而更稳定的生物学结构被保留下来。</span></p><p data-tool="mdnice编辑器"><span leaf="">然后是 query 阶段：scArches 并不会把 reference 和 query 再全部放在一起从头训练一次，而是在已经训练好的 reference model 上做一次很小幅度的更新。对于新的 query 数据，模型只需要为新的 query study labels 增加一小部分新的可训练参数，这些新增参数被称为 </span><strong><span leaf="">adaptors</span></strong><span leaf="">。reference 阶段已经学好的主体参数尽量保持不动，主要训练的只是和新 query 相关的这一小部分权重。</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/Go1sLc9TF8WUR6Tu4NfFu0o3keHVGJOl9iaJ8u5JfNOoZFu7QEASZ6Xg2rRARNJlibCWiccnXvEv0LUtNqfNqgxKIH3ZOwkRQEpp8Nln0nO81w/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5064814814814815" data-type="png" data-w="1080" data-imgfileid="100056498" data-aistatus="1" src="https://mmbiz.qpic.cn/mmbiz_png/Go1sLc9TF8WUR6Tu4NfFu0o3keHVGJOl9iaJ8u5JfNOoZFu7QEASZ6Xg2rRARNJlibCWiccnXvEv0LUtNqfNqgxKIH3ZOwkRQEpp8Nln0nO81w/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-tool="mdnice编辑器"><span><span leaf="">reference model的构建</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">这一部分对应的是 reference atlas 的准备工作。在 </span><code><span leaf="">scvi-tools</span></code><span leaf=""> 的官方 reference mapping tutorial 中，reference 的第一步是训练 </span><code><span leaf="">SCVI</span></code><span leaf="">，用 reference 数据学习一个稳定的 latent space。这个 latent space 后面会作为 query mapping 的参考坐标系。这里需要输入一列条件信息</span><code><span leaf="">batch</span></code><span leaf="">，而且 reference 需要使用原始 counts 作为输入。</span></p><pre data-tool="mdnice编辑器"><code><span leaf="">batch_col = </span><span><span leaf="">"batch"</span></span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">adata_ref.obs[</span><span><span leaf="">"cell_type"</span></span><span leaf="">] = adata_ref.obs[</span><span><span leaf="">"cell_type"</span></span><span leaf="">].cat.add_categories([</span><span><span leaf="">"Unknown"</span></span><span leaf="">])</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">scvi.model.SCVI.setup_anndata(</span><span leaf=""><br></span><span leaf="">    adata_ref,</span><span leaf=""><br></span><span leaf="">    layer=</span><span><span leaf="">None</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    batch_key=batch_col,</span><span leaf=""><br></span><span leaf="">    labels_key=</span><span><span leaf="">"cell_type"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">scvi_model = scvi.model.SCVI(</span><span leaf=""><br></span><span leaf="">    adata_ref,</span><span leaf=""><br></span><span leaf="">    n_latent=</span><span><span leaf="">20</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">scvi_model.train()</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">接下来在预训练好的 </span><code><span leaf="">SCVI</span></code><span leaf=""> 基础上继续构建 </span><code><span leaf="">SCANVI</span></code><span leaf="">：</span></p><pre data-tool="mdnice编辑器"><code><span leaf="">scanvi_model = scvi.model.SCANVI.from_scvi_model(</span><span leaf=""><br></span><span leaf="">    scvi_model,</span><span leaf=""><br></span><span leaf="">    labels_key=</span><span><span leaf="">"cell_type"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    unlabeled_category=</span><span><span leaf="">"Unknown"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">scanvi_model.train(max_epochs=</span><span><span leaf="">20</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">这里还需要把 </span><code><span leaf="">Unknown</span></code><span leaf=""> 加入 reference 的标签类别中，主要是为了给后续无标签 query 预留入口。query 映射进来之后，模型会把它们先当作未标注细胞处理，再在 reference 学到的标签体系里进行预测。</span></p><p data-tool="mdnice编辑器"><span leaf="">保存一下 reference model：</span></p><pre data-tool="mdnice编辑器"><code><span leaf="">scanvi_model.save(</span><span><span leaf="">"reference_scanvi"</span></span><span leaf="">, overwrite=</span><span><span leaf="">True</span></span><span leaf="">, save_anndata=</span><span><span leaf="">True</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span><span leaf="">迁移学习</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">在做 query mapping 之前先运行 </span><code><span leaf="">prepare_query_anndata()</span></code><span leaf="">。这一步的作用是把 query 的特征顺序重排成与 reference 一致，并对 reference 中存在但 query 缺失的特征补零。</span></p><pre data-tool="mdnice编辑器"><code><span leaf="">scvi.model.SCANVI.prepare_query_anndata(adata_query, </span><span><span leaf="">"reference_scanvi"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">query 特征空间对齐后，就可以使用 </span><code><span leaf="">load_query_data()</span></code><span leaf=""> 基于 reference model 构建 query model。</span></p><pre data-tool="mdnice编辑器"><code><span leaf="">scarches_model = scvi.model.SCANVI.load_query_data(</span><span leaf=""><br></span><span leaf="">    adata=adata_query,</span><span leaf=""><br></span><span leaf="">    reference_model=</span><span><span leaf="">"reference_scanvi"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    freeze_dropout=</span><span><span leaf="">True</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">scarches_model.train(max_epochs=</span><span><span leaf="">200</span></span><span leaf="">, plan_kwargs=dict(weight_decay=</span><span><span leaf="">0.0</span></span><span leaf="">))</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">建议在 query 训练阶段将 </span><code><span leaf="">weight_decay</span></code><span leaf=""> 设为 </span><code><span leaf="">0.0</span></code><span leaf="">，这样可以保证 reference 细胞如果通过这个 query model 再计算 latent representation，其表示保持不变。而且因为 reference 训练时使用的 </span><code><span leaf="">batch_key</span></code><span leaf=""> 是 </span><code><span leaf="">batch</span></code><span leaf="">，所以 query 也必须提供同名列。</span></p><p data-tool="mdnice编辑器"><span leaf="">模型训练结束后，首先提取 query 的 latent representation：</span></p><pre data-tool="mdnice编辑器"><code><span leaf="">adata_query.obsm[</span><span><span leaf="">"X_scVI"</span></span><span leaf="">] = scarches_model.get_latent_representation()</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">接下来就可以基于这套表示做邻域图和 UMAP：</span></p><pre data-tool="mdnice编辑器"><code><span leaf="">sc.pp.neighbors(adata_query, use_rep=</span><span><span leaf="">"X_scVI"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">sc.tl.umap(adata_query)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">在自动注释这一步，使用的是 </span><code><span leaf="">SCANVI.predict()</span></code><span leaf="">：</span></p><pre data-tool="mdnice编辑器"><code><span leaf="">adata_query.obs[</span><span><span leaf="">"cell_type_pred"</span></span><span leaf="">] = scarches_model.predict(adata_query)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span><span leaf="">需要注意的点</span></span></h2><ul><li><section><span leaf="">batch_key 的选择非常重要，</span><code><span leaf="">SCVI/SCANVI</span></code><span leaf=""> 会将指定的协变量纳入模型，如果将本应保留的生物学差异错误地作为 nuisance covariate 输入，latent space 的结构就可能受到不必要的影响。如果将本应保留的生物学差异也作为 covariate 输入，就可能出现过校正。</span></section></li><li><section><span leaf="">scArches 流程要求 query 与 reference 使用相同基因集合；如果 query 缺少 reference 基因，就需要补零。论文还指出，随着 reference 与 query 的特征差异增大，性能会下降；因此，reference 构建阶段通常建议保留较大的 HVG 集合，以增加后续 feature overlap 的稳健性。</span></section></li><li><section><span leaf="">scArches 的上限受 reference 质量限制。如果 reference 标签不可靠、覆盖范围不足，或者与 query 在技术平台和生物学背景上差异过大，那么迁移学习和自动注释的可靠性都会受到影响。因此，自动注释结果需要结合 joint embedding、概率分数和 marker genes 一起解释，进行人工复核。</span></section></li></ul></section><section><span leaf=""><br></span></section><p><mp-style-type data-value="10000"></mp-style-type></p></div>
